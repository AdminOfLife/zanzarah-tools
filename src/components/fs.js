import AFRAME from 'aframe/src';

AFRAME.registerComponent('fs', {
  init() {
    const { el: { sceneEl } } = this;
    const { systems: { fs } } = sceneEl;
    this.fs = fs;
  },

  registerFileSystem(fs) {
    this.fs.fileSystems.push(fs);
    return () => {
      const index = this.fileSystems.indexOf(fs);
      if (index === -1) {
        return;
      }
      this.fs.fileSystems.splice(index, 1);
    };
  },
  getFilesystemById(id) {
    return this.fs.fileSystems.find(fs => (
      fs.id === id
    ));
  },
});
