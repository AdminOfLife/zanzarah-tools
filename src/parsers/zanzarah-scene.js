const jBinary = require('jbinary');

const {
  Hex,
  Switch,
  DumpContext,
  DynamicArray,
  DynamicString,
} = require('../utils/jbinary');

const {
  LightTypes,
  TriggerTypes,
  EffectTypes,
} = require('./zanzarah-scene-constants');

exports.typeSet = {
  'jBinary.all': 'Scene',
  'jBinary.littleEndian': true,

  Hex,
  Switch,
  DumpContext,
  DynamicArray,
  DynamicString,

  Scene: jBinary.Template({
    baseType: ['array', 'Section'],
    read() {
      // Remove first and last sections - header and eos
      const sections = this.baseRead().slice(1, -1);
      return sections.reduce((result, section) => {
        result[section.name.slice(1, -1)] = section;
        delete section.name;
        return result;
      }, {});
    },
    write(value) {
      const sectionNames = ['Scenefile'];
      sectionNames.push(...Object.keys(value));
      sectionNames.push('EOS');
      return this.baseWrite(sectionNames.map(name =>
        Object.assign(value[name], {
          name: `[${name}]`,
        })
      ));
    },
  }),

  Section: ['extend', {
    name: 'DynamicString',
  }, ['Switch', context => {
    const handler = `${context.name.slice(1, -1)}Data`;
    if (!exports.typeSet[handler]) {
      throw new Error(`Unsupported section: ${context.name}`);
    }
    return handler;
  }], {
    _: ['DumpContext', null, 1],
  }],

  ScenefileData: {},
  VersionData: {
    author: 'DynamicString',
    _unknown: ['array', 'uint32', 3],
    buildVersion: 'uint32',
    date: 'DynamicString',
    time: 'DynamicString',
    year: 'uint32',
    _unknown1: 'uint32',
  },
  MiscData: ['extend', {
    sceneFile: 'DynamicString',
    scenePath: 'DynamicString',
    texturePath: 'DynamicString',
    _unknownColor: 'FloatColor',
    _unknownVector1: 'Vector3',
    _unknownVector2: 'Vector3',
    _unknownBytes: ['array', 'uint8', 4],
    _flag: 'uint8',
  }, ['if', '_flag', {
    _unknownOptionalBytes: ['array', 'uint8', 4],
    _unknownFloat: 'float32',
  }], {
    _unkownFloats: ['array', 'float32', 2],
  }],
  LightsData: {
    lights: ['DynamicArray', ['extend', {
      id: 'uint32',
      type: ['enum', 'uint32', LightTypes],
      color: 'FloatColor',
      param: 'uint32',
    }, ['if', context => context.type === 'UnknownLight1', {
      _unknownVectors: ['array', 'Vector3', 2],
    }], ['if', context => context.type === 'UnknownLight128', {
      _unknownFloat: 'float32',
      _unkownVector: 'Vector3',
    }], ['if', context => context.type === 'UnknownLight128', {
      _unknownInt: 'uint32',
      _unknownVectors: ['array', 'Vector3', 2],
    }]]],
  },
  FOModels_v4Data: {
    models: ['DynamicArray', {
      id: 'uint32',
      fileName: 'DynamicString',
      _unknownVectors: ['array', 'Vector3', 2],
      _unknownFloats: ['array', 'float32', 5],
      _unknownColor: 'IntColor',
      _unkownFlags: ['array', 'uint8', 2],
      _unkownInt1: 'uint32',
      _unkownFlag: 'uint8',
      _unkownInt2: 'int32',
    }],
  },
  Models_v3Data: {
    models: ['DynamicArray', {
      id: 'uint32',
      fileName: 'DynamicString',
      position: 'Vector3',
      rotation: 'Vector3',
      scale: 'Vector3',
      _unknownColor: 'IntColor',
      _unkownFlag1: 'uint8',
      _unknownInt: 'int32',
      _unkownFlag2: 'uint8',
    }],
  },
  DynamicModelsData: {
    models: ['DynamicArray', {
      id: 'uint32',
      _unkownInts: ['array', 'uint32', 2],
      position: 'Vector3',
      rotation: 'Vector3',
      _unkownFloats: ['array', 'float32', 2],
      _unknownVector: 'Vector3',
      _unknownInts2: ['array', 'uint32', 2],
      _unknownThings: ['array', {
        _unknownFloats: ['array', 'float32', 7],
        _unknownFlag: 'uint8',
        _unknownColor: 'IntColor',
        _unknownInt: 'uint32',
        _unknownString1: 'DynamicString',
        _unknownString2: 'DynamicString',
      }, 3],
    }],
  },
  TriggersData: {
    triggers: ['DynamicArray', ['extend', {
      id: 'uint32',
      type: ['enum', 'uint32', TriggerTypes],
      _unknownFlag: 'uint32',
      _unknownVector: 'Vector3',
      _unknownInts: ['array', 'uint32', 5],
      _unknownString: 'DynamicString',
    }, ['if', context => context.type === 'UnknownTrigger0', {
      _unknownVectors: ['array', 'Vector3', 2],
    }], ['if', context => context.type === 'UnknownTrigger1', {
      _unkownVector: 'Vector3',
      _unknownFloat: 'float32',
    }], ['if', context => context.type === 'UnknownTrigger2', {
      _unkownVector: 'Vector3',
    }]]],
  },
  '2DSamples_v2Data': {
    samples: ['DynamicArray', {
      id: 'uint32',
      fileName: 'DynamicString',
      _unknownInts: ['array', 'uint32', 2],
      _unknownFlag: 'uint8',
    }],
  },
  '3DSamples_v2Data': {
    samples: ['DynamicArray', {
      id: 'uint32',
      fileName: 'DynamicString',
      _unknownVectors: ['array', 'Vector3', 3],
      _unknownInts: ['array', 'uint32', 5],
    }],
  },
  Effects_v2Data: {
    effects: ['DynamicArray', ['extend', {
      id: 'uint32',
      type: ['enum', 'uint32', EffectTypes],
      _unknownInts: ['array', 'uint32', 5],
    }, ['if', context => context.type === 'UnknownEffect1', {
      _unknownInt: 'uint32',
      _unknownVectors: ['array', 'Vector3', 2],
    }], ['if', context => context.type === 'UnknownEffect6', {
      _unknownInt: 'uint32',
      _unknownVectors: ['array', 'Vector3', 2],
    }], ['if', context => context.type === 'UnknownEffect10', {
      _unknownInt: 'uint32',
      _unknownVectors: ['array', 'Vector3', 2],
    }], ['if', context => context.type === 'SnowFlakesEffect', {
      _unknownInt: 'uint32',
    }]]],
  },
  AmbientSoundData: {
    count: 'uint32',
  },
  MusicData: {
    count: 'uint32',
  },
  SceneData: {
    _unknownItems: ['DynamicArray', {
      _unknownString: 'DynamicString',
      _unknownInts: ['array', 'uint32', 2],
    }],
  },
  VertexModifiersData: {
    modifiers: ['DynamicArray', ['extend', {
      id: 'uint32',
      type: 'uint32',
      _unknownVector: 'Vector3',
      _unknownColor: 'IntColor',
    }, ['if', context => context.type === 1, {
      _unknownVector2: 'Vector3',
    }], {
      _unknownInt: 'uint32',
      _unknownFlag: 'uint8',
    }]],
  },
  BehavioursData: {
    behaviours: ['DynamicArray', {
      type: 'uint32',
      modelId: 'uint32',
    }],
  },
  DatasetData: {
    _unknownData: ['DynamicArray', 'uint8'],
    _unknownStrings: ['array', 'DynamicString', 2],
  },
  SceneOriginData: {
    origin: 'Vector3',
  },
  TexturePropertiesData: {
    properties: ['DynamicArray', {
      fileName: 'DynamicString',
      value: 'int32',
    }],
  },
  WaypointSystemData: ['extend', {
    version: 'uint32',
    empty: 'uint32',
  }, ['if', context => context.version >= 5, {
    _unknownData: ['array', 'uint8', 24],
    _hack: ['array', 'uint32', 2],
  }], {
    _reserved: ['const', 'uint32', 0xFFFF, true],
  }],
  BackdropData: {
    fileName: 'DynamicString',
  },
  EOSData: {},

  Vector3: ['array', 'float32', 3],
  FloatColor: ['array', 'float32', 4],
  IntColor: ['Hex', 'uint32', 8],
};
