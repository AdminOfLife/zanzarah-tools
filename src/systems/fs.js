import AFRAME from 'aframe/src';
import _debug from 'debug';

const info = _debug('app:systems:fs:info');

AFRAME.registerSystem('fs', {
  schema: {},

  init() {
    this.fileSystems = [];
  },

  async load(options, onLoad, onProgress, onError) {
    for (const fs of this.fileSystems) {
      const hasFile = await fs.includes(options);
      if (!hasFile) {
        continue;
      }
      info('Using fs %s to fulfil the request %o', fs.id, options);
      return fs.load(options, onLoad, onProgress, onError);
    }
    info('File not found %s', options.url);
    if (onError) {
      onError(new Error(`File not found: ${options.url}`));
    }
  },
  async index() {
    const allFiles = [];
    for (const fs of this.fileSystems) {
      allFiles.push(...await fs.index());
    }
    return allFiles;
  },
});
