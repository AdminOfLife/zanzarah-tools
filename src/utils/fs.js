import RemoteFS from './filesystems/RemoteFS';
import { getRootPath } from './paths';

export default new RemoteFS(getRootPath());
