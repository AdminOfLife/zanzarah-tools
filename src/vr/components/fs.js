import AFRAME from 'aframe/src';

AFRAME.registerComponent('fs', {
  init() {
    const { el: { sceneEl } } = this;
    const { systems: { fs } } = sceneEl;
    this.fsSystem = fs;
  },

  addFileSystem(fs) {
    return this.fsSystem.current.add(fs);
  },
});
