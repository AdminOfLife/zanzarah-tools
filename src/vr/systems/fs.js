import AFRAME from 'aframe/src';

import CompositeFS from 'utils/filesystems/CompositeFS';

AFRAME.registerSystem('fs', {
  init() {
    this.current = new CompositeFS();
  },
});
