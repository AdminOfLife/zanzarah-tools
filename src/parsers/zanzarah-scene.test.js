import fs from 'fs';
import jBinary from 'jbinary';
import { typeSet } from './zanzarah-scene';

const parse = async fileNameOrBuffer => {
  let binary = null;
  if (Buffer.isBuffer(fileNameOrBuffer)) {
    binary = new jBinary(fileNameOrBuffer, typeSet);
  } else {
    binary = await jBinary.load(fileNameOrBuffer, typeSet);
  }
  return binary.readAll();
};

describe('parsers - zanzarah scene', () => {
  it('exports typeSet to parse', () => {
    expect(typeSet).toBeDefined();
  });
  pit('parses empty file', async () => {
    expect(await parse('./public/fixtures/empty.dff')).toEqual({});
  });
  pit('parses smallest known files', async () => {
    expect(await parse('./public/fixtures/smallest-scene.scn')).toMatchSnapshot();
  });
});
