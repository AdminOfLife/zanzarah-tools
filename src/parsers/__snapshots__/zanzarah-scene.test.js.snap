exports[`parsers - zanzarah scene parses smallest known files 1`] = `
Object {
  "2DSamples_v2": Object {
    "samples": Array [],
  },
  "3DSamples_v2": Object {
    "samples": Array [],
  },
  "AmbientSound": Object {
    "count": 0,
  },
  "Backdrop": Object {
    "fileName": "5_Swamp",
  },
  "Behaviours": Object {
    "behaviours": Array [
      Object {
        "modelId": 1,
        "type": 900,
      },
    ],
  },
  "Dataset": Object {
    "_unknownData": Array [
      181,
      12,
      0,
      0,
      0,
      0,
      0,
      0,
      177,
      62,
      117,
      192,
      0,
      0,
      205,
      205,
      1,
      0,
      0,
      0,
      0,
      205,
      205,
      205,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    "_unknownStrings": Array [
      "",
      "",
    ],
  },
  "DynamicModels": Object {
    "models": Array [],
  },
  "Effects_v2": Object {
    "effects": Array [],
  },
  "FOModels_v4": Object {
    "models": Array [
      Object {
        "_unknownColor": "ffffffff",
        "_unknownFloats": Array [
          8,
          12,
          1,
          1,
          1,
        ],
        "_unknownVectors": Array [
          Array [
            10.557332038879395,
            31.438854217529297,
            13.713237762451172,
          ],
          Array [
            0.3493209779262543,
            0,
            -0.937002420425415,
          ],
        ],
        "_unkownFlag": 0,
        "_unkownFlags": Array [
          0,
          0,
        ],
        "_unkownInt1": 0,
        "_unkownInt2": -1,
        "fileName": "ifl04h",
        "id": 28,
      },
    ],
  },
  "Lights": Object {
    "lights": Array [],
  },
  "Misc": Object {
    "_flag": 1,
    "_unknownBytes": Array [
      57,
      48,
      36,
      0,
    ],
    "_unknownColor": Array [
      0,
      0,
      0,
      1,
    ],
    "_unknownFloat": 0,
    "_unknownOptionalBytes": Array [
      41,
      52,
      31,
      0,
    ],
    "_unknownVector1": Array [
      8.091316223144531,
      31.32491683959961,
      9.48287582397461,
    ],
    "_unknownVector2": Array [
      0.6837635040283203,
      -0.5698028802871704,
      0.4558423161506653,
    ],
    "_unkownFloats": Array [
      0.10000000149011612,
      46,
    ],
    "sceneFile": "sc_3253",
    "scenePath": "..\\Resources\\Worlds\\",
    "texturePath": "..\\Resources\\Textures\\Worlds\\",
  },
  "Models_v3": Object {
    "models": Array [
      Object {
        "_unknownColor": "ff868686",
        "_unknownInt": -1,
        "_unkownFlag1": 0,
        "_unkownFlag2": 0,
        "fileName": "ido01h",
        "id": 1,
        "position": Array [
          11.674039840698242,
          31.57732391357422,
          16.910022735595703,
        ],
        "rotation": Array [
          -0.7141005396842957,
          0,
          0.7000429630279541,
        ],
        "scale": Array [
          1,
          1,
          1,
        ],
      },
      Object {
        "_unknownColor": "ffffffff",
        "_unknownInt": -1,
        "_unkownFlag1": 0,
        "_unkownFlag2": 1,
        "fileName": "ifl03h",
        "id": 2,
        "position": Array [
          10.511805534362793,
          31.437274932861328,
          13.36729907989502,
        ],
        "rotation": Array [
          0.03404809534549713,
          0,
          -0.9994198083877563,
        ],
        "scale": Array [
          1,
          1,
          1,
        ],
      },
    ],
  },
  "Music": Object {
    "count": 3452816845,
  },
  "Scene": Object {
    "_unknownItems": Array [
      Object {
        "_unknownInts": Array [
          28,
          2,
        ],
        "_unknownString": "",
      },
      Object {
        "_unknownInts": Array [
          1,
          3,
        ],
        "_unknownString": "",
      },
      Object {
        "_unknownInts": Array [
          15,
          4,
        ],
        "_unknownString": "doorway 1243",
      },
      Object {
        "_unknownInts": Array [
          16,
          4,
        ],
        "_unknownString": "single player",
      },
      Object {
        "_unknownInts": Array [
          17,
          4,
        ],
        "_unknownString": "ns3104",
      },
      Object {
        "_unknownInts": Array [
          19,
          4,
        ],
        "_unknownString": "",
      },
      Object {
        "_unknownInts": Array [
          22,
          4,
        ],
        "_unknownString": "",
      },
      Object {
        "_unknownInts": Array [
          23,
          4,
        ],
        "_unknownString": "ns3105",
      },
      Object {
        "_unknownInts": Array [
          24,
          4,
        ],
        "_unknownString": "",
      },
      Object {
        "_unknownInts": Array [
          2,
          3,
        ],
        "_unknownString": "",
      },
      Object {
        "_unknownInts": Array [
          25,
          4,
        ],
        "_unknownString": "",
      },
    ],
  },
  "SceneOrigin": Object {
    "origin": Array [
      14.091316223144531,
      26.32491683959961,
      13.48287582397461,
    ],
  },
  "TextureProperties": Object {
    "properties": Array [
      Object {
        "fileName": "default",
        "value": -2,
      },
      Object {
        "fileName": "ifl00th",
        "value": 14,
      },
      Object {
        "fileName": "ifl03th",
        "value": 14,
      },
      Object {
        "fileName": "xsp01th",
        "value": 14,
      },
      Object {
        "fileName": "xsp04th",
        "value": 14,
      },
      Object {
        "fileName": "xsp06th",
        "value": 14,
      },
      Object {
        "fileName": "xwo00th",
        "value": 14,
      },
      Object {
        "fileName": "xwo09th",
        "value": 14,
      },
    ],
  },
  "Triggers": Object {
    "triggers": Array [
      Object {
        "_unknownFlag": 0,
        "_unknownFloat": 1.149999976158142,
        "_unknownInts": Array [
          0,
          13,
          13,
          1243,
          0,
        ],
        "_unknownString": "",
        "_unknownVector": Array [
          0.7591406106948853,
          -0.05777226760983467,
          -0.648405909538269,
        ],
        "_unkownVector": Array [
          10.371908187866211,
          31.653139114379883,
          17.129343032836914,
        ],
        "id": 15,
        "type": "UnknownTrigger1",
      },
      Object {
        "_unknownFlag": 0,
        "_unknownInts": Array [
          1,
          0,
          0,
          0,
          0,
        ],
        "_unknownString": "",
        "_unknownVector": Array [
          -0.17345939576625824,
          -0.1488700956106186,
          -0.9735873341560364,
        ],
        "_unkownVector": Array [
          11.955608367919922,
          30.400392532348633,
          15.556184768676758,
        ],
        "id": 16,
        "type": "UnknownTrigger2",
      },
      Object {
        "_unknownFlag": 0,
        "_unknownInts": Array [
          3,
          2207288756,
          0,
          0,
          0,
        ],
        "_unknownString": "",
        "_unknownVector": Array [
          -0.745530366897583,
          -0.5600088834762573,
          -0.3613913953304291,
        ],
        "_unkownVector": Array [
          13.085081100463867,
          30.560749053955078,
          13.454812049865723,
        ],
        "id": 17,
        "type": "UnknownTrigger2",
      },
      Object {
        "_unknownFlag": 0,
        "_unknownInts": Array [
          5,
          1,
          2,
          0,
          0,
        ],
        "_unknownString": "",
        "_unknownVector": Array [
          -0.1186022087931633,
          -0.9817796945571899,
          -0.14863041043281555,
        ],
        "_unkownVector": Array [
          12.872611999511719,
          30.1561279296875,
          13.440271377563477,
        ],
        "id": 19,
        "type": "UnknownTrigger2",
      },
      Object {
        "_unknownFlag": 0,
        "_unknownInts": Array [
          5,
          2,
          2,
          0,
          0,
        ],
        "_unknownString": "",
        "_unknownVector": Array [
          0.8578603267669678,
          -0.47725433111190796,
          0.19066941738128662,
        ],
        "_unkownVector": Array [
          14.100581169128418,
          30.1541805267334,
          15.156375885009766,
        ],
        "id": 22,
        "type": "UnknownTrigger2",
      },
      Object {
        "_unknownFlag": 0,
        "_unknownInts": Array [
          3,
          1567274756,
          0,
          0,
          0,
        ],
        "_unknownString": "",
        "_unknownVector": Array [
          -0.846588671207428,
          -0.40055859088897705,
          0.350474089384079,
        ],
        "_unkownVector": Array [
          16.19293785095215,
          30.24066162109375,
          11.976113319396973,
        ],
        "id": 23,
        "type": "UnknownTrigger2",
      },
      Object {
        "_unknownFlag": 0,
        "_unknownInts": Array [
          5,
          0,
          0,
          0,
          0,
        ],
        "_unknownString": "",
        "_unknownVector": Array [
          -0.24567367136478424,
          -0.9686687588691711,
          -0.03636462613940239,
        ],
        "_unkownVector": Array [
          13.876931190490723,
          30.25874137878418,
          13.780603408813477,
        ],
        "id": 24,
        "type": "UnknownTrigger2",
      },
      Object {
        "_unknownFlag": 0,
        "_unknownInts": Array [
          5,
          3,
          2,
          0,
          0,
        ],
        "_unknownString": "",
        "_unknownVector": Array [
          -0.09985959529876709,
          -0.9138188362121582,
          0.39349132776260376,
        ],
        "_unkownVector": Array [
          13.260506629943848,
          30.234582901000977,
          14.429807662963867,
        ],
        "id": 25,
        "type": "UnknownTrigger2",
      },
    ],
  },
  "Version": Object {
    "_unknown": Array [
      0,
      0,
      0,
    ],
    "_unknown1": 3,
    "author": "snoopy",
    "buildVersion": 4,
    "date": "13.11.2001",
    "time": "13:05",
    "year": 2001,
  },
  "VertexModifiers": Object {
    "modifiers": Array [],
  },
  "WaypointSystem": Object {
    "_hack": Array [
      0,
      0,
    ],
    "_reserved": 65535,
    "_unknownData": Array [
      0,
      112,
      190,
      77,
      60,
      104,
      59,
      163,
      114,
      17,
      209,
      61,
      0,
      142,
      175,
      55,
      16,
      67,
      32,
      1,
      12,
      124,
      18,
      0,
    ],
    "empty": 0,
    "version": 6,
  },
}
`;
